mappings/net/minecraft/entity/Entity.mapping

```diff
@@ -192,13 +223,26 @@ CLASS ail net/minecraft/entity/Entity
 		ARG 2 target
 	METHOD a setBoundingBox (Lcrv;)V
 		ARG 1 boundingBox
+	METHOD a applyPistonMovement (Lcsa;)Lcsa;
+		ARG 1 offset
 	METHOD a movementInputToVelocity (Lcsa;FF)Lcsa;
 		ARG 0 movementInput
 		ARG 1 speed
 		ARG 2 yaw
 	METHOD a clipSneakingMovement (Lcsa;Laiz;)Lcsa;
 		ARG 1 offset
 		ARG 2 type
+	METHOD a calculatePerpendicularMotionVector (Lcsa;Lcrv;Lbho;Lail;)Lcsa;
+		ARG 0 collisionVector
+		ARG 1 observerBounds
+		ARG 2 world
+		ARG 3 subject
+	METHOD a calculateTangentialMotionVector (Lcsa;Lcrv;Lbho;Lcsf;Lzx;)Lcsa;
+		ARG 0 collisionVector
+		ARG 1 observerBounds
+		ARG 2 world
+		ARG 3 context
+		ARG 4 collisions
```

Comment:
These were named wrong. They have nothing to do with perpendicular and tangential motion. Both adjust a movement for collisions, but one is an optimized version that can only be used when the movement is along a single axis (walking up stairs calls the collision movement adjustment method many times with movement along only one axis):

```java
    public static Vec3d adjustMovementForCollisions(@Nullable Entity entity, Vec3d movement, Box entityBoundingBox, World world, EntityContext context, ReusableStream<VoxelShape> collisions) {
        boolean noX = movement.x == 0.0;
        boolean noY = movement.y == 0.0;
        boolean noZ = movement.z == 0.0;

        if (noX && noY || noX && noZ || noY && noZ) {
            return adjustSingleAxisMovementForCollisions(movement, entityBoundingBox, world, context, collisions);
        }

        ReusableStream<VoxelShape> blockCollisions = new ReusableStream<>(Stream.concat(collisions.stream(), world.getBlockCollisions(entity, entityBoundingBox.stretch(movement))));
        return adjustMovementForCollisions(movement, entityBoundingBox, blockCollisions);
    }
```

Reply:
Since this has already been merged why don't you make a new PR/issue so we can continue discussion?

Reply:
I found these issues while working on a new PR, and the PR will include changes to these methods.